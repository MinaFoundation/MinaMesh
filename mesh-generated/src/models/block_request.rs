/*
 * Rosetta
 *
 * Build Once. Integrate Your Blockchain Everywhere. 
 *
 * The version of the OpenAPI document: 1.4.13
 * 
 * Generated by: https://openapi-generator.tech
 */

use crate::models;
use serde::{Deserialize, Serialize};

/// BlockRequest : A BlockRequest is utilized to make a block request on the /block endpoint. 
#[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
pub struct BlockRequest {
    #[serde(rename = "network_identifier")]
    pub network_identifier: Box<models::NetworkIdentifier>,
    #[serde(rename = "block_identifier")]
    pub block_identifier: Box<models::PartialBlockIdentifier>,
}

impl BlockRequest {
    /// A BlockRequest is utilized to make a block request on the /block endpoint. 
    pub fn new(network_identifier: models::NetworkIdentifier, block_identifier: models::PartialBlockIdentifier) -> BlockRequest {
        BlockRequest {
            network_identifier: Box::new(network_identifier),
            block_identifier: Box::new(block_identifier),
        }
    }
}

