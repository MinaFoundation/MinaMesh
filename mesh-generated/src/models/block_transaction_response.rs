/*
 * Rosetta
 *
 * Build Once. Integrate Your Blockchain Everywhere. 
 *
 * The version of the OpenAPI document: 1.4.13
 * 
 * Generated by: https://openapi-generator.tech
 */

use crate::models;
use serde::{Deserialize, Serialize};

/// BlockTransactionResponse : A BlockTransactionResponse contains information about a block transaction. 
#[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
pub struct BlockTransactionResponse {
    #[serde(rename = "transaction")]
    pub transaction: Box<models::Transaction>,
}

impl BlockTransactionResponse {
    /// A BlockTransactionResponse contains information about a block transaction. 
    pub fn new(transaction: models::Transaction) -> BlockTransactionResponse {
        BlockTransactionResponse {
            transaction: Box::new(transaction),
        }
    }
}

