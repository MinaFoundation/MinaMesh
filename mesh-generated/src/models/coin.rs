/*
 * Rosetta
 *
 * Build Once. Integrate Your Blockchain Everywhere. 
 *
 * The version of the OpenAPI document: 1.4.13
 * 
 * Generated by: https://openapi-generator.tech
 */

use crate::models;
use serde::{Deserialize, Serialize};

/// Coin : Coin contains its unique identifier and the amount it represents. 
#[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
pub struct Coin {
    #[serde(rename = "coin_identifier")]
    pub coin_identifier: Box<models::CoinIdentifier>,
    #[serde(rename = "amount")]
    pub amount: Box<models::Amount>,
}

impl Coin {
    /// Coin contains its unique identifier and the amount it represents. 
    pub fn new(coin_identifier: models::CoinIdentifier, amount: models::Amount) -> Coin {
        Coin {
            coin_identifier: Box::new(coin_identifier),
            amount: Box::new(amount),
        }
    }
}

