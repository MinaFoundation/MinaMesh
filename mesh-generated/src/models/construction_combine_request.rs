/*
 * Rosetta
 *
 * Build Once. Integrate Your Blockchain Everywhere. 
 *
 * The version of the OpenAPI document: 1.4.13
 * 
 * Generated by: https://openapi-generator.tech
 */

use crate::models;
use serde::{Deserialize, Serialize};

/// ConstructionCombineRequest : ConstructionCombineRequest is the input to the `/construction/combine` endpoint. It contains the unsigned transaction blob returned by `/construction/payloads` and all required signatures to create a network transaction. 
#[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
pub struct ConstructionCombineRequest {
    #[serde(rename = "network_identifier")]
    pub network_identifier: Box<models::NetworkIdentifier>,
    #[serde(rename = "unsigned_transaction")]
    pub unsigned_transaction: String,
    #[serde(rename = "signatures")]
    pub signatures: Vec<models::Signature>,
}

impl ConstructionCombineRequest {
    /// ConstructionCombineRequest is the input to the `/construction/combine` endpoint. It contains the unsigned transaction blob returned by `/construction/payloads` and all required signatures to create a network transaction. 
    pub fn new(network_identifier: models::NetworkIdentifier, unsigned_transaction: String, signatures: Vec<models::Signature>) -> ConstructionCombineRequest {
        ConstructionCombineRequest {
            network_identifier: Box::new(network_identifier),
            unsigned_transaction,
            signatures,
        }
    }
}

