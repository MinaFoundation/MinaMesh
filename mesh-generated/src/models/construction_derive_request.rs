/*
 * Rosetta
 *
 * Build Once. Integrate Your Blockchain Everywhere. 
 *
 * The version of the OpenAPI document: 1.4.13
 * 
 * Generated by: https://openapi-generator.tech
 */

use crate::models;
use serde::{Deserialize, Serialize};

/// ConstructionDeriveRequest : ConstructionDeriveRequest is passed to the `/construction/derive` endpoint. Network is provided in the request because some blockchains have different address formats for different networks. Metadata is provided in the request because some blockchains allow for multiple address types (i.e. different address for validators vs normal accounts). 
#[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
pub struct ConstructionDeriveRequest {
    #[serde(rename = "network_identifier")]
    pub network_identifier: Box<models::NetworkIdentifier>,
    #[serde(rename = "public_key")]
    pub public_key: Box<models::PublicKey>,
    #[serde(rename = "metadata", skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Value>,
}

impl ConstructionDeriveRequest {
    /// ConstructionDeriveRequest is passed to the `/construction/derive` endpoint. Network is provided in the request because some blockchains have different address formats for different networks. Metadata is provided in the request because some blockchains allow for multiple address types (i.e. different address for validators vs normal accounts). 
    pub fn new(network_identifier: models::NetworkIdentifier, public_key: models::PublicKey) -> ConstructionDeriveRequest {
        ConstructionDeriveRequest {
            network_identifier: Box::new(network_identifier),
            public_key: Box::new(public_key),
            metadata: None,
        }
    }
}

