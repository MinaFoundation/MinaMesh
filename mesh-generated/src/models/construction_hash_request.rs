/*
 * Rosetta
 *
 * Build Once. Integrate Your Blockchain Everywhere. 
 *
 * The version of the OpenAPI document: 1.4.13
 * 
 * Generated by: https://openapi-generator.tech
 */

use crate::models;
use serde::{Deserialize, Serialize};

/// ConstructionHashRequest : ConstructionHashRequest is the input to the `/construction/hash` endpoint. 
#[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
pub struct ConstructionHashRequest {
    #[serde(rename = "network_identifier")]
    pub network_identifier: Box<models::NetworkIdentifier>,
    #[serde(rename = "signed_transaction")]
    pub signed_transaction: String,
}

impl ConstructionHashRequest {
    /// ConstructionHashRequest is the input to the `/construction/hash` endpoint. 
    pub fn new(network_identifier: models::NetworkIdentifier, signed_transaction: String) -> ConstructionHashRequest {
        ConstructionHashRequest {
            network_identifier: Box::new(network_identifier),
            signed_transaction,
        }
    }
}

