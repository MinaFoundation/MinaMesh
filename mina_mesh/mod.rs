pub mod account;
mod block;
pub mod construction;
mod mempool;
pub mod network;

use crate::graphql_generated::mina::QueryNetworkId;
use anyhow::{bail, Context as AnyhowContext, Result};
use cynic::{http::ReqwestExt, QueryBuilder};
use mesh::models::NetworkIdentifier;
use reqwest::Client;
use serde::Deserialize;
use sqlx::PgPool;
use std::vec::Vec;

#[derive(Deserialize, Debug, Default)]
pub struct Config {
  #[serde(default = "default_mina_proxy_url")]
  mina_proxy_url: String,
  #[serde(default = "default_archive_url")]
  archive_url: String,
  #[serde(default = "default_database_url")]
  database_url: String,
  genesis_block_identifier: String,
}

fn default_mina_proxy_url() -> String {
  "https://mainnet.minaprotocol.network/graphql".to_string()
}

fn default_archive_url() -> String {
  "https://api.minascan.io/archive/devnet/v1/graphql".to_string()
}

fn default_database_url() -> String {
  "postgres://mina:whatever@localhost:5432/archive".to_string()
}

pub struct MinaMeshContext {
  config: Config,
  client: Client,
  pool: PgPool,
}

impl MinaMeshContext {
  async fn from_env() -> Result<Self> {
    let config = envy::from_env::<Config>().with_context(|| "Failed to parse config from env")?;
    let database_url = config.database_url.clone();
    Ok(Self {
      config,
      client: Client::new(),
      pool: PgPool::connect(database_url.as_str()).await?,
    })
  }

  async fn network_health_check(&self, network_identifier: NetworkIdentifier) -> Result<bool> {
    let QueryNetworkId { network_id } = self.graphql(QueryNetworkId::build(())).await?;
    if network_identifier.blockchain == "MINA" {
      unimplemented!();
    }
    if &network_identifier.network == &network_id {
      unimplemented!();
    }
    Ok(true)
  }

  async fn graphql<ResponseData, Vars>(&self, operation: cynic::Operation<ResponseData, Vars>) -> Result<ResponseData>
  where
    Vars: serde::Serialize,
    ResponseData: serde::de::DeserializeOwned + 'static,
  {
    let response = self
      .client
      .post(&self.config.mina_proxy_url)
      .run_graphql(operation)
      .await
      .context("Failed to run GraphQL query")?;
    if let Some(errors) = response.errors {
      bail!(errors
        .iter()
        .map(|err| err.message.clone())
        .collect::<Vec<String>>()
        .join("\n\n"));
    } else if let Some(data) = response.data {
      Ok(data)
    } else {
      bail!("No data contained in GraphQL response");
    }
  }
}

trait ToVecOfString {
  fn to_vec_of_string(self) -> Vec<String>;
}

impl ToVecOfString for Vec<&str> {
  fn to_vec_of_string(self) -> Vec<String> {
    self.into_iter().map(|s| s.to_string()).collect()
  }
}
