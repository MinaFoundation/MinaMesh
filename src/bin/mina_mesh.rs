// TODO: document workflow regarding fetching and using initial genesis ledger
// hash.

use anyhow::Result;
use clap::Parser;
use mina_mesh::{FetchGenesisBlockIdentifierCommand, ServeCommand};

#[derive(Debug, Parser)]
#[command(
  name = "mina-mesh",
  version,
  about = "A Mesh-compliant Server for Mina",
  propagate_version = true,
  author
)]
enum Command {
  #[command(about = "Start the Mina Mesh Server.")]
  Serve(ServeCommand),
  #[command(about = "Retrieve the genesis block identifier via a proxy node GraphQL endpoint.")]
  FetchGenesisBlockIdentifier(FetchGenesisBlockIdentifierCommand),
}

#[tokio::main]
async fn main() -> Result<()> {
  match Command::parse() {
    Command::Serve(cmd) => cmd.execute().await,
    Command::FetchGenesisBlockIdentifier(cmd) => cmd.execute().await,
  }
}
