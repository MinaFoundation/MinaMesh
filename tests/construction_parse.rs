use anyhow::Result;
use coinbase_mesh::models::ConstructionParseRequest;
use insta::assert_debug_snapshot;
use mina_mesh::{
  test::{
    network_id, signed_transaction_delegation, signed_transaction_payment, unsigned_transaction_delegation,
    unsigned_transaction_payment,
  },
  MinaMeshConfig,
};

#[tokio::test]
async fn construction_parse_valid_signed_payment() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionParseRequest {
    network_identifier: network_id().into(),
    transaction: signed_transaction_payment(),
    signed: true,
  };

  let response = mina_mesh.construction_parse(request).await;
  assert!(response.is_ok());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_parse_valid_signed_delegation() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionParseRequest {
    network_identifier: network_id().into(),
    transaction: signed_transaction_delegation(),
    signed: true,
  };

  let response = mina_mesh.construction_parse(request).await;
  assert!(response.is_ok());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_parse_valid_unsigned_payment() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionParseRequest {
    network_identifier: network_id().into(),
    transaction: unsigned_transaction_payment(),
    signed: false,
  };

  let response = mina_mesh.construction_parse(request).await;
  assert!(response.is_ok());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_parse_valid_unsigned_delegation() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionParseRequest {
    network_identifier: network_id().into(),
    transaction: unsigned_transaction_delegation(),
    signed: false,
  };

  let response = mina_mesh.construction_parse(request).await;
  assert!(response.is_ok());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_parse_invalid_fee() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionParseRequest {
    network_identifier: network_id().into(),
    transaction: r#"{
            "signature": "EE1D10B5EF283026177B8C61F75C84F09B35C94C6D1417C2C88707E2D26CBB21D1371F16F3AEC696E055C235D9EA2F707630EB395813AEBFE120BBDD5B5E8908",
            "payment": {
                "to": "B62qm3B76ruDJc4aQJyEw3iTjpHJ9xrTceJosqVFkhoUTbSjEdF3xgU",
                "from": "B62qogmsu15o7DJQkzLtAux6THLrfPeLUJLzpmiBPkEWaYLdWze4Tj9",
                "fee": "10", 
                "token": "1",
                "nonce": "10",
                "memo": "test payment",
                "amount": "100",
                "valid_until": "200000"
            },
            "stake_delegation": null
        }"#
      .to_string(),
    signed: true,
  };

  let response = mina_mesh.construction_parse(request).await;
  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_parse_invalid_format_signed() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionParseRequest {
    network_identifier: network_id().into(),
    transaction: r#"{
            "signature": "EE1D10B5EF283026177B8C61F75C84F09B35C94C6D1417C2C88707E2D26CBB21D1371F16F3AEC696E055C235D9EA2F707630EB395813AEBFE120BBDD5B5E8908",
            "payment": null,
            "stake_delegation": null  
        }"#
      .to_string(),
    signed: true,
  };

  let response = mina_mesh.construction_parse(request).await;
  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_parse_invalid_format_unsigned() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionParseRequest {
    network_identifier: network_id().into(),
    transaction: r#"{
            "randomOracleInput": "000000033024262EC0279FAE8D8984B4DFF694E3DE1A8513B8DB9E13A2BA190581EC974A3024262EC0279FAE8D8984B4DFF694E3DE1A8513B8DB9E13A2BA190581EC974A5E6737A0AC0A147918437FC8C21EA57CECFB613E711CA2E4FD328401657C291C000002570561800000000000800000000000000081F0000001586000407C0382808784808087848080A1C2A1C287848080C30607848080C10784808086000E0000000000000000013E815200000000",
            "signerInput": { "prefix": [], "suffix": [] },
            "payment": null,
            "stakeDelegation": null  
        }"#
      .to_string(),
    signed: false,
  };

  let response = mina_mesh.construction_parse(request).await;
  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_parse_invalid_operations_signed() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionParseRequest {
    network_identifier: network_id().into(),
    transaction: r#"{
            "signature": "FE1D10B5EF283026177B8C61F75C84F09B35C94C6D1417C2C88707E2D26CBB21D1371F16F3AEC696E055C235D9EA2F707630EB395813AEBFE120BBDD5B5E8908",
            "payment": {
                "to": "B62qm21Bx7bV1mGvk5JwBy3QjxTY71MNg6FQNNR6ce4HEHoUGsKRcYA",
                "from": "B62qm21Bx7bV1mGvk5JwBy3QjxTY71MNg6FQNNR6ce4HEHoUGsKRcYQ",
                "fee": "100000000",
                "token": "1",
                "nonce": "10",
                "memo": "test payment",
                "amount": "100",
                "valid_until": "200000"
            },
            "stake_delegation": null
        }"#
      .to_string(),
    signed: true,
  };

  let response = mina_mesh.construction_parse(request).await;
  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}

#[tokio::test]
async fn construction_parse_invalid_operations_unsigned() -> Result<()> {
  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;

  let request = ConstructionParseRequest {
    network_identifier: network_id().into(),
    transaction: r#"{
      "randomOracleInput":"000000031BAED3545AE7F7236ED78A3B7F059768D7C6E9A039A512AFF346F1C51F8661941BAED3545AE7F7236ED78A3B7F059768D7C6E9A039A512AFF346F1C51F8661947F3BA0154F672B1E82F84C207B84502A38A8967D61C86727E43054C0FFB59026000002570242F000000000008000000000000000080000000021EBE840500B531B1B7B00000000000000000000000000000000000000000000000000000006000000000000000005A4640000000000",
      "signerInput":{
          "prefix":[
              "1BAED3545AE7F7236ED78A3B7F059768D7C6E9A039A512AFF346F1C51F866194",
              "1BAED3545AE7F7236ED78A3B7F059768D7C6E9A039A512AFF346F1C51F866194",
              "7F3BA0154F672B1E82F84C207B84502A38A8967D61C86727E43054C0FFB59026"
          ],
          "suffix":[
              "01BDB1B195A014042FAF080000000020000000000000000200000000001E8480",
              "0000000003000000000000000000000000000000000000000000000000000000",
              "0000000000000000000000000000000000000000000000000004C4B400000000"
          ]
      },
      "payment":{
          "to":"B62qj7nR7j5GiQLJEBrMq49nX8fKcLJKK57kDdja7w9YPJQMdsshtcL",
          "from":"B62qqEMfUYCW4ePTDE9ZGVfn42ugxQ6CSe8PdTCrRpZAnCRezAGsVp7",
          "fee":"1000000",
          "token":"wSHV2S4qX9jFsLjQo8r1BsMLH2ZRKsZx6EJd1sbozGPieEC4Jf",
          "nonce":"8",
          "memo":"hello",
          "amount":"10777000000",
          "valid_until":"200000000"
      },
      "stakeDelegation":null
    }"#.to_string(),
    signed: false,
  };

  let response = mina_mesh.construction_parse(request).await;
  assert!(response.is_err());
  assert_debug_snapshot!(response);
  Ok(())
}
