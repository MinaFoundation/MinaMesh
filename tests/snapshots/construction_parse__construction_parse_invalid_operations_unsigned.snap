---
source: tests/construction_parse.rs
expression: response
---
Err(
    JsonParse(
        Some(
            "Unsigned transaction does not match operations, randomOracleInput or signerInput",
        ),
    ),
)
